### PR TITLE
feat: add unit argument and apply conversion for kms

### LINKED ISSUE #338: Haversine distance macro doesn't support kms
### Describe the bug
The macro to calculate the haversine distance between two points always returns its result in miles, which isn't useful for people who use the metric system

### Steps to reproduce
- Work somewhere other than [Liberia, Myanmar, the United Kingdom and the United States](https://en.wikipedia.org/wiki/Mile#International_mile)
- Use the haversine_distance macro

### Expected results
- The option to get results in either miles or kilometres (or even kilometers)

### Actual results
- Results in miles

### System information
**The contents of your `packages.yml` file:**
```yaml
packages:
  - package: fishtown-analytics/dbt_utils
    version: 0.6.3
  
  - package: fishtown-analytics/audit_helper
    version: 0.3.0

  - package: fishtown-analytics/codegen
    version: 0.3.1

  - package: fishtown-analytics/redshift
    version: 0.4.1
```
**Which database are you using dbt with?**
- [ ] postgres
- [x] redshift
- [ ] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
0.19.0 (dbt Cloud)
```

### Additional context
The [great circle calculator](https://andrew.hedges.name/experiments/haversine/) linked from the original macro's source says to use 6373km as the radius of the Earth, which is as good a number as any

### Are you interested in contributing the fix?
Yes, but not for a while.

### PR DESCRIPTION
**Pull request by bastienboutonnet**
This is a:
- [x] bug fix PR with no breaking changes — please ensure the base branch is `master`
- [x] new functionality — please ensure the base branch is the latest `dev/` branch
**To be honest I'm not too sure what this counts as. #338  is labelled as a bug, but it's also kind of a feature request --let me know what you want to consider this one as and I'll adjust the diff target**

- [ ] a breaking change — please ensure the base branch is the latest `dev/` branch

## Description & motivation
the `haversine_distance()` calculation outputs results in miles only at the minute. Interest in being able to output haversine in kilometers was registered in #338 .

## Changes
calling with `unit='km'` (distance between Amsterdam and Paris)
```sql
select {{haversine_distance(48.864716, 2.349014, 52.379189, 4.899431, unit='km')}} as hav_dist
from table_a
```
returns ~ 430
Screenshot 2021-02-27 at 14 58 53

calling witout a unit param or (`unit='mi'`) 

```sql
select {{haversine_distance(48.864716, 2.349014, 52.379189, 4.899431)}} as hav_dist
from table_a
```
returns ~267
Screenshot 2021-02-27 at 14 57 34

Closes #338 

## Checklist
- [x] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [ ] BigQuery
    - [x] Postgres
    - [ ] Redshift
    - [ ] Snowflake
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)

- [ ] I have added an entry to CHANGELOG.md
    - Not sure if this change would cause a minor or patch increment. In semver world since this add a backwards compatible features it should trigger a bump to `0.7.0` but not sure if you follow that here, so let me know and then I can update the changelog accordingly.

### CODE REVIEW

**Comment by joellabes** [comment]
🙌 Thanks @bastienboutonnet! 

This repo contains another project which imports dbt-utils and runs integration tests against seed files. Check out https://github.com/fishtown-analytics/dbt-utils/pull/334/files for an example of configuring tests

**Comment by bastienboutonnet** [comment]
Thanks for the pointers @joellabes I'll see about writing a test for this function at some point.

**Comment by bastienboutonnet** [comment]
@joellabes I added some tests and they seem to pass locally and on circle CI except for snowflake and BQ. I'm assuming because I'm a 3rd party these tests are bound to fail? Either way, let me know if that's the issue of if there is anything I should do (or not do) in the code.

Regarding the changelog todo on the PR template my question still stands let me know what I should do when we get closer to merge I guess?

**Comment by joellabes** [comment]
> I'm assuming because I'm a 3rd party these tests are bound to fail?

I don't _think_ that's true - I've had all the adapters succeed on past PRs. My best guess is that BQ and SF do _something_ differently - perhaps their column type inference isn't playing nicely with the seed, or the `conversion_rate` multiplication doesn't work for some reason. I only have access to a Redshift warehouse so can't muck around to find out why, and I'm external to Fishtown too so I can't see the CircleCI errors - @clrcrl will have to let you know the specifics. 

In `integration_tests/models/geo/test_haversine_distance_mi.sql`, it might be nice to union that model and run it again without passing in the `mi` argument to test the default/backwards-compatible case as well (unless you have that covered somewhere else that I missed). 

0.7.0 is the next significant (intentionally avoiding semver words) version of dbt-utils, so I expect that this sort of net-new functionality would go there as opposed to a patch version. Again a good q for Claire

**Comment by bastienboutonnet** [comment]
@joellabes with the union part of your message I'm not sure I understand. I think I understand you'd like a test where we call the macro without the uni argument. But I don't know why union. Did you have something in mind?

**Comment by joellabes** [comment]
> Did you have something in mind?

Broadly, I was thinking something like 

```
select convert(lat_1, lon_1, lat_2, lon_2, ‘mi’)
from table

union all

select convert(lat_1, lon_1, lat_2, lon_2)
from table
```

(written on my phone so only very broadly accurate)

By doing that, you’d get 0, 1 or 2 failing rows depending on which part broke. It’s probably bad practice to overload the test, but I was trying to avoid having to make a third separate query 😬  

not sure if there’s a SF-friendly version of the macro sorry! Weird that it got this far without breaking. Maybe it’s never worked but no one uses this on BQ or SF

**Comment by bastienboutonnet** [comment]
@joellabes Ok I see what you meant. I can get that done in that way I think it's not that bad actually coming to think about it.

So yeah I checked into my team's macro for haversine and we have used the `pow()` functions on our Snowflake warehouse. I think there is a way to distribute warehouse specific macros and let the macro adaptor choose it at runtime, I'll see if there is an example of this already on the repo or wait till you or @clrcrl let me know how best to do it.

But YES! very strange people didn't complain so far, or maybe they have been like us and have used their own macros for haversine. Anyway, would be a great idea to fix that for everyone right?

**Comment by bastienboutonnet** [comment]
@joellabes I just implemented the unioned test and it seems to work fine on pg and redshift.

What do you think should happen qua the other dbs?

**Comment by bastienboutonnet** [comment]
I think I may just be good with creating a new macro and name it `snowflake__haversine_distance` with the db specific function in it and dbt would take care of selecting it based on the adapter type?

For BQ I was looking and they don't seem to have `radians` function to convert something that's in degrees + I cannot test it as I can't play with a BQ warehouse. Looks like if we implemented it for BQ we'd also have to make some maths happen to convert radians to degrees. It's probably doable though.

Let me know what makes sense.

**Comment by joellabes** [comment]
Sorry for the delay getting back to you! 

> I think I may just be good with creating a new macro and name it `snowflake__haversine_distance` with the db specific function in it and dbt would take care of selecting it based on the adapter type?

Yes I think that's right! 

Re BigQuery, if it was already broken and you don't use BQ, I don't think you need to feel obliged to bend over backwards to make it work. You could either add an implementation that no-ops and returns null, or throw the equivalent of a [NotImplementedException](https://docs.microsoft.com/en-us/dotnet/api/system.notimplementedexception?view=net-5.0), or do nothing at all and it will throw a more confusing error. When a BQ user comes along who wants this, they can always open another PR. I _think_ there are other macros that do this but I don't remember which offhand.

**Comment by bastienboutonnet** [comment]
no worries at all! 

Ok I'll have a go at a snowflake implementation for sure because I can actually test it out. For BQ I think what you propose makes sense, however not sure why you're pointing me to the Microsoft website here:

>  or throw the equivalent of a NotImplementedException,

I'd actually like to be able to throw a proper error if BQ users were to use this macro rather than a null. How best do you think we can do that?

**Comment by joellabes** [comment]
> not sure why you're pointing me to the Microsoft website here

In a past life I was a C# developer, so that was the concept I knew it as - I'm not sure what the dbtonic way of doing this is. 

The closest equivalent I could find was https://github.com/fishtown-analytics/dbt-utils/blob/c64b52068070023bc3524525e06a14ad72d485a4/macros/cross_db_utils/datediff.sql#L55. I think you'd basically have to make a stub of a bigquery implementation that threw the exception. 

Then there would have to be some way of excluding that one from CI, because an exception that's intentionally thrown is still an exception... Hopefully @clrcrl has some way of excluding them somehow 🤔

**Comment by bastienboutonnet** [comment]
@joellabes so I looked into the postgres and redshift documentations and they also have the `pow()` function that is the only way to make snowflake calculate power. So in the interest of not duplicating too much code I went ahead and made the `default__` implementation use `pow()`. Judging by the tests this seems to work fine for PG, redshift and snowflake which is really nice.

I'll work on making the BQ exception raising happen soon.

**Comment by bastienboutonnet** [comment]
@joellabes @clrcrl in the end I went ahead and also worked an adaptor macro for big query, it was a bit of a pain but I managed. I created so many failing CI runs I hope you folks don't get notified on each failures.

Anyway, let me know if you have any additional feedback

**Comment by joellabes** [comment]
Incredible, thanks so much @bastienboutonnet! Sorry that this turned out to be so much more work than first thought 😭

**Comment by bastienboutonnet** [comment]
Ah no worries, if I didn't want to do it I would have said it. I was stuck on my other open source projects today so it was something fun and useful to do.

Is there a changelog entry I should be writing btw? And or some documentation?

Also, not sure if this should be merged into the main branch or develop. 

I guess by this point this PR is more of a feature than a bugfix so I rebase on `dev/` and change the target of the PR.

**Comment by joellabes** [comment]
Not sure on those logistics - Claire's got "clear the packages backlog" on her sprint for this fortnight so I imagine she'll be checking this out pretty soon though!

File: macros/geo/haversine_distance.sql [hunk 1]
```diff
@@ -3,17 +3,48 @@ This calculates the distance between two sets of latitude and longitude.
...
+{% macro degrees_to_radians(degrees) -%}
+    acos(-1) * {{degrees}} / 180
+{%- endmacro %}
+
+{% macro haversine_distance(lat1,lon1,lat2,lon2,unit='mi') -%}
+    {{ return(adapter.dispatch('haversine_distance', packages = dbt_utils._get_utils_namespaces())(lat1,lon1,lat2,lon2,unit)) }}
 {% endmacro %}
 
-{% macro default__haversine_distance(lat1,lon1,lat2,lon2) -%}
+{% macro default__haversine_distance(lat1,lon1,lat2,lon2,unit) -%}
```
Comment (clrcrl) [inline macros/geo/haversine_distance.sql:17]:
```suggestion
{% macro default__haversine_distance(lat1, lon1, lat2, lon2, unit='mi') -%}
```

File: macros/geo/haversine_distance.sql [hunk 2]
```diff
@@ -3,17 +3,48 @@ This calculates the distance between two sets of latitude and longitude.
...
+{# vanilla macro is in miles #}
+    {% set conversion_rate = '' %}
```
Comment (clrcrl) [inline macros/geo/haversine_distance.sql:19]: This is a little cleaner, and has some error handling!
```suggestion
{%- if unit == 'mi' %}
   {% set conversion_rate = 1 %}
{% elif unit == 'km' %}
   {% set conversion_rate = 1.60934 %}
{% else %}
    {{ exceptions.raise_compiler_error("unit input must be one of 'mi' or 'km'. Got ' ~ unit) }}
{% endif -%}
```

File: macros/geo/haversine_distance.sql [hunk 3]
```diff
@@ -3,17 +3,48 @@ This calculates the distance between two sets of latitude and longitude.
...
+{% if unit == 'km' %}
+{# we multiply miles result to get it in kms #}
+    {% set conversion_rate = '* 1.60934' %}
+{% endif %}
 
-    2 * 3961 * asin(sqrt((sin(radians(({{lat2}} - {{lat1}}) / 2))) ^ 2 +
+    2 * 3961 * asin(sqrt(pow((sin(radians(({{lat2}} - {{lat1}}) / 2))), 2) +
     cos(radians({{lat1}})) * cos(radians({{lat2}})) *
-    (sin(radians(({{lon2}} - {{lon1}}) / 2))) ^ 2))
+    pow((sin(radians(({{lon2}} - {{lon1}}) / 2))), 2))) {{conversion_rate}}
```
Comment (clrcrl) [inline macros/geo/haversine_distance.sql:27]:
```suggestion
    pow((sin(radians(({{ lon2 }} - {{ lon1 }}) / 2))), 2))) * {{ conversion_rate }}
```

File: macros/geo/haversine_distance.sql [hunk 4]
```diff
@@ -3,17 +3,48 @@ This calculates the distance between two sets of latitude and longitude.
...
 {%- endmacro %}
+
+
+
+{% macro bigquery__haversine_distance(lat1,lon1,lat2,lon2,unit) -%}
+{% set radians_lat1 = dbt_utils.degrees_to_radians(lat1) %}
+{% set radians_lat2 = dbt_utils.degrees_to_radians(lat2) %}
+{% set radians_lon1 = dbt_utils.degrees_to_radians(lon1) %}
+{% set radians_lon2 = dbt_utils.degrees_to_radians(lon2) %}
+{# vanilla macro is in miles #}
```
Comment (clrcrl) [inline macros/geo/haversine_distance.sql:38]: same as above!

**Review by clrcrl — CHANGES_REQUESTED** [review]
Love these changes! Yay km!!!

I wrote some suggestions for how this code might be made a little cleaner :)

File: macros/geo/haversine_distance.sql (hunk 1 above)
Comment (bastienboutonnet) [reply macros/geo/haversine_distance.sql:17]: done, wasn't sure if we wanted to implement it on the specific macros but that makes sense.

File: macros/geo/haversine_distance.sql (hunk 2 above)
Comment (bastienboutonnet) [reply macros/geo/haversine_distance.sql]: Thanks thats a lot better indeed! Done :)

**Comment by bastienboutonnet** [comment]
Hey @clrcrl thanks a lot for the feedback all done. Hope this works 🥳

**Review by clrcrl — APPROVED** [review]
🎉

**Comment by bastienboutonnet** [comment]
@clrcrl Thanks for merging the PR. Btw, I had not added any changelog entries yet because I wasn't sure whether this was bug or feature. Would you like me to add something to the changelog or you will do it when you release?

**Pull request merged**